### ISSUE TITLE
How to speed up the calculation of func '_min_norm_2d'?

### ISSUE THREAD

**Issue by zeng-hello-world** [intel-isl/MultiObjectiveOptimization#8]
I found that if the `vecs ` input into func `_min_norm_2d` is too large, the whole procedure becomes very slow. In my experiment, the vecs.shape = [39321600,1], and one single training iteration costs 40s, while the `_min_norm_2d` spends 35s+.
The experiment in your paper assume that the share layers shape is much smaller mine, so I guess maybe this is the point.

**Comment by ozansener** [comment]
Which version are you using, numpy or pytorch? If it is pytorch min_norm_2d should not be that slow. Regardless, I think there is another issue in your case. `vecs` should be a list of vectors such that each entry is a gradient. So, it is basically [grad_of_loss1, grad_of_loss2, ...] etc. 

If the grad_of_loss1 is [39321600,1] you are a bit in trouble since you need to compute dot products of [39321600,1] size vectors (Line 8 in Alg 2 in the paper). The best you can do in my opinion is create Mp = [NumTasks, 39321600,1] matrix and compute M = torch.mm(Mp.transpose, Mp) if the number of tasks is really large. You can benchmark it and see which one is faster.
